File: 01-matrix/01-matrix.java
import java.util.*;

class Solution {
    public int[][] updateMatrix(int[][] mat) {
        int numRows = mat.length, numCols = mat[0].length;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        ArrayDeque<int[]> deque = new ArrayDeque<int[]>();
        boolean[][] visited = new boolean[numRows][numCols];
        int[][] result = new int[numRows][numCols];

        // Add 0s to the deque
        for (int row = 0; row < numRows; row++) {
            for (int col = 0; col < numCols; col++) {
                if (mat[row][col] == 0) {
                    deque.addLast(new int[]{row, col});
                    visited[row][col] = true; 
                }
            }
        }

        int distance = 0;
        while (deque.size() > 0) {
            int breadth = deque.size();

            while (breadth > 0) {
                int[] curr = deque.removeFirst();
                int row = curr[0], col = curr[1];

                visited[row][col] = true; 
                result[row][col] = distance; 
                
                // Add neighbours to deque
                for (int[] dir : directions) {
                    int nRow = row + dir[0], nCol = col + dir[1];
                    boolean isValid = (0 <= nRow) && (nRow < numRows) && (0 <= nCol) && (nCol < numCols);

                    if (isValid && !visited[nRow][nCol]) {
                        deque.addLast(new int[]{nRow, nCol});
                        visited[nRow][nCol] = true; 

                    }
                }
                breadth -= 1;
            }
            distance += 1;
        }

        return result; 
    }
}

File: 01-matrix/01-matrix.py
from collections import deque
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        numRows, numColumns = len(mat), len(mat[0])
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        result = [([0] * numColumns) for _ in range(numRows)]
        queue = deque()
        visited = set()

        # Add Os to the queue 
        for row in range(numRows):
            for col in range(numColumns):
                if (mat[row][col] == 0):
                    queue.append((row, col))
                    visited.add((row, col))

        distance = 0 
        while (len(queue) > 0):
            for _ in range(len(queue)):
                row, col = queue.popleft()

                result[row][col] = distance

                if (result[row][col] == "1"):
                    minDistance = min(distance, minDistance)

                for d in directions:
                    nRow, nCol = row + d[0], col + d[1]

                    isValid = (0 <= nRow < numRows) and (0 <= nCol < numColumns)
                    if (isValid and (nRow, nCol) not in visited):
                        queue.append((nRow, nCol))
                        visited.add((nRow, nCol))

            distance += 1

        return (result)


            

